File: app/src/main/java/com/monk/rxjava2/imetate2/ObservableCreator.java
package com.monk.rxjava2.imetate2;

import java.util.concurrent.atomic.AtomicReference;

/**
 * @author monk
 * @date 2019-03-01
 */
public class ObservableCreator<T> extends Observable<T> {
    private ObservableOnSubscribe<T> observableOnSubscribe;

    public ObservableCreator(ObservableOnSubscribe<T> observableOnSubscribe) {
        this.observableOnSubscribe = observableOnSubscribe;
    }

    @Override
    public void subscribeActual(Observer<T> observer) {
        ObservableEmitter<T> observableEmitter = new ObservableEmitter<>(observer);
        observer.onSubscribe(observableEmitter);
        // 最关键的调用
        observableOnSubscribe.subscribe(observableEmitter);
    }

    public static final class ObservableEmitter<T> extends AtomicReference<Disposable> implements Emitter<T>, Disposable {
        private Observer<T> observer;

        public ObservableEmitter(Observer<T> observer) {
            this.observer = observer;
        }

        @Override
        public boolean isDisposed() {
            return DisposableHelper.isDisposed(get());
        }

        @Override
        public void dispose() {
            DisposableHelper.dispose(this);
        }

        @Override
        public void onNext(T t) {
            if (!isDisposed()) {
                observer.onNext(t);
            }
        }

        @Override
        public void onCompleted() {
            if (!isDisposed()) {
                observer.onCompleted();
            }
        }

        @Override
        public void onError(Throwable throwable) {
            if (!isDisposed()) {
                observer.onError(throwable);
            }
        }
    }
}

File: module_fragment/src/main/java/com/monk/modulefragment/rxjava2/imetate2/ObservableCreator.kt
package com.monk.modulefragment.rxjava2.imetate2

import java.util.concurrent.atomic.AtomicReference

/**
 * @author monk
 * @date 2019-03-01
 */
class ObservableCreator<T>(private val observableOnSubscribe: ObservableOnSubscribe<T>) : Observable<T>() {
    override fun subscribeActual(observer: Observer<T>?) {
        val observableEmitter = ObservableEmitter(observer)
        observer!!.onSubscribe(observableEmitter)
        // 最关键的调用
        observableOnSubscribe.subscribe(observableEmitter)
    }

    class ObservableEmitter<T>(private val observer: Observer<T>?) : AtomicReference<Disposable?>(), Emitter<T>, Disposable {
        override fun isDisposed(): Boolean {
            return DisposableHelper.isDisposed(get()!!)
        }

        override fun dispose() {
            DisposableHelper.dispose(this)
        }

        override fun onNext(t: T) {
            if (!isDisposed()) {
                observer!!.onNext(t)
            }
        }

        override fun onCompleted() {
            if (!isDisposed()) {
                observer!!.onCompleted()
            }
        }

        override fun onError(throwable: Throwable?) {
            if (!isDisposed()) {
                observer!!.onError(throwable)
            }
        }
    }
}